Md3 migration

Bnyro opened the pull request:
Hey,
this PR is an attempt to completely migrate the app to Material Design 3.
If available, dynamic colors are being used, otherwise the default red accent colors are being chosen instead.
![1](https://user-images.githubusercontent.com/82752168/228476293-cc635598-6f71-4f0f-86aa-97700b071252.png)
![2](https://user-images.githubusercontent.com/82752168/228476298-fc19b3ce-df54-44a7-8be7-41c121aa3b35.png)
![3](https://user-images.githubusercontent.com/82752168/228476303-df5740f0-f303-4a3c-9fa5-9b7a7c422107.png)
![4](https://user-images.githubusercontent.com/82752168/228476314-0afc2dbe-8df2-48bd-a45d-0f1009ca43cf.png)


closes #3390

PerfectSlayer commented:
Hi @Bnyro 

Nice update you did there! I had not the chance/time to look at MD3 yet.
Is there any issue or missing part you are aware of with the port?

It feels like the homescreen would need some love again, at least for consistency.
Round badges with icons (blocked, allowed, redirected, log, help and support) do no more match the rest of the UI.

I could start a new branch so we can iterate on this.
What do you think of it?

Bnyro commented:
Hey @PerfectSlayer,
could you please point that out? You mean an issue with the preference for it or something?

I've changed the design of the rounded badges a bit within the last few commits but forgot to update the screenshot actually, it should fit way better now. I don't think breaking changes are needed since we probably also want to keep the user experience consistent.

However, if there are some suggestions I don't mind working on implementations on a different branch.

PS: I'll update the screenshots later once I find time for it.

PerfectSlayer commented:
Don't worry about updating the screenshot, I can test it on my side.
Overall, it's a great change. There is some minor issues I think, mainly with older device (like the Pixel 2 I used to test).

The first thing I notice is the purple button from the homescreen bottom bar (maybe a default style/value?):
![Screenshot_20230401-083140](https://user-images.githubusercontent.com/1766222/229269954-57fabfc9-ec5b-43d0-ba57-b80e88a49c3f.png)

The update and support pages now feel `dull` compared to the strong white on red, especially their controls:

![Screenshot_20230401-083119](https://user-images.githubusercontent.com/1766222/229270051-aa1cfc05-85fa-435c-956d-fe22ed5a850e.png)
![Screenshot_20230401-083149](https://user-images.githubusercontent.com/1766222/229270035-8a133e27-fd4b-4bcc-9d9c-e659205c5ea8.png)

![Screenshot_20230401-083153](https://user-images.githubusercontent.com/1766222/229270070-fa70ee8b-1ddb-4f80-8783-0a7008ae3bef.png)
![Screenshot_20230401-083124](https://user-images.githubusercontent.com/1766222/229270066-dbc1159d-b373-4208-9cbb-801ad26b98a3.png)

Whereas the settings pages now have a more contrasted switch controls:

![Screenshot_20230401-083224](https://user-images.githubusercontent.com/1766222/229270231-c795b4ca-5ba8-492f-b3f2-66c2689f0d6d.png)
![Screenshot_20230401-083227](https://user-images.githubusercontent.com/1766222/229270233-3d23c948-a0f2-42a8-a18a-9b65c44911d1.png)

It should be minor changes and you already did the bigger part of the task.
I will start reading more about MD3 migration to be sure to try to fix thoses parts and be sure I can maintain it in the future 😉

Bnyro commented:
I might have some ideas concerning the issue, did I get you correctly that you want to attempt solving these on your own? Otherwise I don't mind working on fixing these, but anyways I'm always open for it if you have any questions concerning Material Design 3 as I do have quite a lot of experience with it! So don't hesitate to ask! :)

PerfectSlayer commented:
> did I get you correctly that you want to attempt solving these on your own? 

Not really. I will have to do it in order to migrate to MD3 but I don't care who fix it 😅 
The think I want is to have a global understanding of what I merged to the project. Not do everything on my own 😄 

> Otherwise I don't mind working on fixing these, but anyways I'm always open for it if you have any questions concerning Material Design 3 as I do have quite a lot of experience with it! So don't hesitate to ask! :)

Thanks! I start reading at the migration guide and it looks like I will also have to rework `styles.xml` and `colors.xml` as they are using deprecated values like `primaryVariant`... Any thought about it?

Bnyro commented:
If I was you I would use the [`Material Theme Builder`](https://m3.material.io/theme-builder#/dynamic) to generate a new theme based on the currently used primary color.
I didn't want to do that in my PR since I wasn't sure whether accent color changes are welcome, but that would also fix the two issues mentioned above!

PerfectSlayer commented on app/src/main/res/values/styles.xml:
```diff
@@ -1,41 +1,87 @@
+        <item name="colorSurfaceInverse">@color/inverseSurface</item>
+        <item name="colorPrimaryInverse">@color/inversePrimary</item>
 
         <!--Remap legacy AppCompat attributes to MaterialComponent attributes-->
         <item name="colorPrimaryDark">?colorPrimary</item>
```
Is it still needed?

Bnyro replied:
No, we can remove that!

PerfectSlayer commented on app/src/main/AndroidManifest.xml:
```diff
@@ -106,7 +106,7 @@
             android:label="@string/support_title"
             android:parentActivityName=".ui.home.HomeActivity"
-            android:theme="@style/Theme.AdAway.NoActionBar.Red" />
+            android:theme="@style/Theme.AdAway" />
```
Is this needed if you change it to the default theme?

Bnyro replied:
Oh true, that's not needed anymore, silly me.

PerfectSlayer commented on app/src/main/java/org/adaway/ui/prefs/BasePreferenceFragment.java:
```diff
@@ -0,0 +1,66 @@
+    public void onCreatePreferences(@Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState, @Nullable @org.jetbrains.annotations.Nullable String rootKey) {}
+
+    @Override
+    public void onDisplayPreferenceDialog(@NonNull Preference preference) {
+        if (preference instanceof ListPreference) {
+            showListPreference((ListPreference) preference);
+        } else if (preference instanceof EditTextPreference) {
+            showEditTextPreference((EditTextPreference) preference);
+        } else {
+            super.onDisplayPreferenceDialog(preference);
+        }
+    }
```
Is this required to use MD components by default?

Bnyro replied:
Yes, that's needed for the preferences to look material 3 designed.
Preferences are not yet officially available at Material Design 3, so we have to override the default Material Design 2 preference dialogs and show our own custom, MaterialAlertDialogBuilder instances instead, which use MD3.

PerfectSlayer commented on app/src/main/res/values/styles.xml:
```diff
@@ -1,41 +1,87 @@
-        <item name="android:statusBarColor">?colorOnSecondary</item>
-        <item name="android:navigationBarColor">@color/welcomeBackground</item>
-        <item name="android:windowLightStatusBar">false</item>
+    <style name="MaterialPreferenceSwitch" parent="@style/Preference.SwitchPreferenceCompat.Material">
+        <item name="widgetLayout">@layout/view_preference_switch</item>
     </style>
```
Can you detail what the `MaterialPreferenceSwitch` style use for? 🙏

Bnyro replied:
That's needed as the preferences are not styled Material Design 3 yet.
By using our custom switch style, our switches look like https://m3.material.io/components/switch/overview instead of the old switches from MD2 (https://m2.material.io/components/switches).

PerfectSlayer reviewed (commented):
That looks way better with it! 👍  What was the settings (initial color?) you used in the Theme Builder?
I will keep testing it to check if I can find remaining quirks but I can't find it on first look.
That's a good bunch of changes but it feels like to move in the right direction for the app 😉

Bnyro commented:
> That looks way better with it! +1 What was the settings (initial color?) you used in the Theme Builder? I will keep testing it to check if I can find remaining quirks but I can't find it on first look. That's a good bunch of changes but it feels like to move in the right direction for the app wink
>
It's been the old primary color, I've been using that as the base color for the new theme (you can find it as the `seed` color in the new themes: #ef5350.

PerfectSlayer commented:
Hey, just a comment to keep you updated.
I have a lot on my plate (the past weeks as the coming weeks) so I would apologize to not getting any progress lately on this PR.
It is definitely a great contribution that goes in the right direction.
So be sure I would like to get back at it as soon as I get more bandwidth 😃 👍

Bnyro commented:
No worries, I know that situation from myself sometimes.
Just take your time, the PR will be waiting for you ;)
From using my branch the last weeks, I've also found one or two other minor appearance issues with the migration that I'll be addressing soon.
Have a good one!

Bnyro commented:
(I've now also added dividers below the Checkbox list items to fit the style of Material Design 3 and fixed some other issues as well as removing the two things you mentioned in your review above)

originalsouth commented:
@Vankog does not seem active, hope he is doing well. @PerfectSlayer please advise on how to proceed?

PerfectSlayer commented:
Hey @cyb3rko, I got your message from my mail. Thanks 🙏 

I think that between you and @Bnyro you should be able to pull something out. What do you need from me to unblock you?
Can't promise anything as I have been under the wave for the past months, but we won't get any progress if I don't let room for others to contribute and support them in the way.

cyb3rko commented:
Hi @PerfectSlayer,  
no problem, I'm always happy to support projects like this, even as a poor student :)

Thanks for the @, haven't seen this PR yet. It's definitely what I'm looking.  
I will take a look at what to improve when I find some time.

cyb3rko commented:
I've commited some fixes: https://github.com/cyb3rko/AdAway/tree/md3-migration
- Shrink home screen icons to not overlap with text
- Fix M2 switch design in app list
- Replace button cardviews with actual M3 buttons
- Remove duplicate attributes and reset card radius to default

Take a look if you want and tell me, what you like and what you do not.

cyb3rko commented:
And what I've noticed is that some resource files are unused and most dialogs still need an overhaul, those using an EditText.

Bnyro commented:
Hey @cyb3rko, thanks for your efforts! From what you described the changes sound reasonable :+1:

Unfortunately I currently am really low in free time (and it probably won't get better soon), so I think it would be awesome if you could just make a new PR at this repository containing my changes from this PR and yours - so that you can also react to feedback of others directly without me being in between.

Cheers
Bnyro

cyb3rko commented:
Hey @Bnyro,  
will do that soon, thanks!

Greetings
Niko

cyb3rko commented:
My changes are to found at https://github.com/AdAway/AdAway/pull/4083 now
